src/org/opensolaris/opengrok/analysis/pascal/PascalXref.lex

```diff
@@ -0,0 +1,162 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").  
+ * You may not use this file except in compliance with the License.
+ *
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2006, 2016, Oracle and/or its affiliates. All rights reserved.
+ */
+
+/*
+ * Cross reference a Pascal file
+ */
+
+package org.opensolaris.opengrok.analysis.pascal;
+import org.opensolaris.opengrok.analysis.JFlexXref;
+import java.io.IOException;
+import java.io.Writer;
+import java.io.Reader;
+import org.opensolaris.opengrok.web.Util;
+
+%%
+%public
+%class PascalXref
+%extends JFlexXref
+%unicode
+%ignorecase
+%int
+%{
+  /* Must match WhiteSpace regex */
+  private final static String WHITE_SPACE = "[ \t\f\r]+";
+
+  // TODO move this into an include file when bug #16053 is fixed
+  @Override
+  protected int getLineNumber() { return yyline; }
+  @Override
+  protected void setLineNumber(int x) { yyline = x; }
+%}
+
+/* Must match WHITE_SPACE constant */
+WhiteSpace     = [ \t\f]+
+EOL = \r|\n|\r\n
+Identifier = [a-zA-Z_] [a-zA-Z0-9_]+
+
+URIChar = [\?\+\%\&\:\/\.\@\_\;\=\$\,\-\!\~\*\\]
+FNameChar = [a-zA-Z0-9_\-\.]
+File = [a-zA-Z]{FNameChar}* "." ("pas"|"properties"|"props"|"xml"|"conf"|"txt"|"htm"|"html"|"ini"|"diff"|"patch")
+Path = "/"? [a-zA-Z]{FNameChar}* ("/" [a-zA-Z]{FNameChar}*[a-zA-Z0-9])+
+
+Number = (0[xX][0-9a-fA-F]+|[0-9]+\.[0-9]+|[0-9]+)(([eE][+-]?[0-9]+)?[ufdlUFDL]*)?
+
+%state  STRING COMMENT SCOMMENT QSTRING
+
+%%
+<YYINITIAL>{
+ "begin"     { incScope(); writeUnicodeChar(yycharat(0)); writeUnicodeChar(yycharat(1));writeUnicodeChar(yycharat(2)); writeUnicodeChar(yycharat(3));writeUnicodeChar(yycharat(4)); }
```

Comment:
At first sight this should work just fine for the scopes. Though I'm not sure about one line functions / procedures... (if there is such a thing in Pascal, I haven't used it for a couple of decades now).

I think you don't need to use writeUnicodeChar(...) since "begin" and "end" does not contain any unicode characters. out.write(yytext()); should be enough.